black-cape/cast-iron-worker
# Patch/improve parallelism

## Linked MR : 

https://github.com/black-cape/cast-iron/pull/4

## Some Cast Iron/Kafka background as relevant to the MR

Cast Iron listens for S3 bucket notifications for mainly two types of files currently, ETL processing configuration in TOML format, and ETL data files to be processed according to the configs.

Cast Iron uses Faust, a stream processing framework similar to [KStream](https://kafka.apache.org/documentation/streams/) which works on Kafka.

Kafka has two consumption modes, controlled by how a consumer group is setup up
* Pub/sub, in which all consumers have different consumer group
* Load balanced, in which all consumers have the SAME consumer group

In Faust parlance, the consumer group is configured in what is called a Faust App, as the Faust App ID.  A Faust Worker can contain one or more Faust App, which App has a different App ID.  This is documented somewhat obscurely [here](https://faust.readthedocs.io/en/latest/userguide/settings.html#std:setting-id)


Kafka Topic is partitioned, one of available Faust Worker is chosen as the leader and assigned that partition.

In Cast Iron, each instance of Cast Iron container we spin up is one Faust Worker. 

In Summary, in a load balanced mode, we need a multiple partitioned Kafka Topic, and Faust Apps running configured with different App ID (Consumer Group)

## Challenge
* Want to be able to scale data processing by spinning up multiple Cast Iron container instances.

## The Challenge a bit more spelled out

Previously one ONE of multiple Cast Iron container instances spun up will process data because only ONE instance gets all the TOML based ETL processing configurations.  This was due to the fact both the ETL configuration and ETL data files were getting pumped through the same topic and one single Faust App configured to run against it the same way. 

That single Faust App had the same consumer group, hence everything was in load balanced consumption mode. 

Cast Iron instance caches the ETL processing configs as a local dictionary, however when operated under load balanced mode each Cast Iron instance would only get a partial set of configs.  

Solution
* Separate handling of TOML based ETL processing configuration FROM data files, leveraging Kafka Topic partitioning and Faust configuration to achieve what we want

* The change allows each container instance to hold on to same cache of ETL processing configurations, while the triggering of ETL on incoming data files are load balanced across multiple cast iron instances.  

* Configure MINIO bucket notifications differently between ETL processing config and ETL source data leveraging bucket notification filters.  Note the filter is a bit [limiting](https://docs.min.io/docs/minio-bucket-notification-guide.html) as there is only include filter, there is no Exclude filter. 

* This is demonstrated in below diagram
    * the ETL processing configuration follows the red path (as known as Pub/Sub in Kafka Consumption mode)
    * The data files follow the black path (as known as load balanced Kafka Consumption mode)

![image](https://user-images.githubusercontent.com/7915744/125644362-8a2e52ba-9e87-4c7a-985a-38f3224c3228.png)


## Other Things also fixed in this MR

* Dumping of error message due to processing error, previously any processing error was not flushed out to temp file on disk before moving over to MINIO

![image](https://user-images.githubusercontent.com/7915744/125644639-e131d2ab-203f-4b19-a056-a6b89c787c70.png)

## Testing Instruction

Please see the MR [here](https://github.com/black-cape/cast-iron/pull/4) for testing instructions
